### DISCUSSION TITLE
Problema na consulta do XML em relação ao recibo

### DISCUSSION THREAD

**Discussion by viniciusfeld** [qualitaocupacional/libesocial#10] [General]
Olá, estou fazendo o uso da lib para enviar eventos do tipo 1200, 1210 e 3000, e tem me facilitado muito, porém me deparei com um erro relacionado ao número do recibo, que não esta sendo reconhecido, e com isso estou recebendo o mesmo retorno do E-social sempre. Consegue me auxiliar em como resolver este problema?

No anexo está um print do retorno que tenho recebido após o envio!

![image](https://github.com/qualitaocupacional/libesocial/assets/72530694/068662ec-f862-45cc-a64f-21827cfd0216)

No outro anexo está o print do arquivo XML que estou enviando, e nele consta o número do recibo.

![image](https://github.com/qualitaocupacional/libesocial/assets/72530694/934e8fd3-624c-401b-8805-1a45f2393705)

Desde já, agradeço e fico no aguardo do retorno.

**Comment by andidevel** [comment]
Olá @viniciusfeld, tudo certo?

Não sei se entendi direito, mas pelos prints parece que você está enviando um evento de retificação (indRetif=2) e está recebendo uma resposta do eSocial dizendo que o seu envio está com algum problema. Quando o retorno é inválido, nenhum recibo é gerado, como você pode perceber que ```recibo: null```.

No primeiro print é possível ver que o retorno do envio "está inválido", pois o código da resposta geral foi "401". Geralmente quando a resposta é **401**, vai ter outras tags de erro explicando qual o problema. Se possível, mostre o XML de retorno ou o dicionário de resposta da LIBeSocial (removendo os dados sensíveis, claro, como CPFs, CNPJs, etc).

Ao que tudo indica, você está retificando um evento de remuneração, informando o código do recibo do evento a ser retificado (na tag **nrRecibo**). OK, esse é o procedimento, mas o eSocial está respondendo que tem algum problema com o seu envio.

O que talvez esteja confundindo você, que também me confundiu no início, é que o eSocial retorna a mensagem "Lote processado com sucesso", o que leva a crer que o seu evento foi processado com sucesso. Na verdade essa mensagem significa que o eSocial processou o seu lote (que pode ter mais de um evento) sem nenhum erro por parte dele, seja de comunicação, banco de dados etc. Mas a "resposta desse processamento com sucesso" é que algum evento do seu lote está com algum problema, e para esse evento não foi gerado um recibo, pois ele não foi aceito.

Se eu entendi errado, por favor informe mais detalhes de quais os procedimentos que você está usando. Se puder mostrar o XML de retorno, sem os dados sensíveis, iria ajudar também.

Espero ter ajudado.

**Reply by viniciusfeld** [reply]
Olá @andidevel, obrigado pelo retorno, foi muito esclarecedor!
Vou te mandar o arquivos XML que eu crio e envio via LIBeSocial, se puder dar uma analisada se estou fazendo da maneira correta a criação desse arquivo e o uso dele dentro da lib. 

OBS: Os dados sensiveis estão com valores aleatorios por motivos de segurança.

[envio_remun.zip](https://github.com/qualitaocupacional/libesocial/files/12090548/envio_remun.zip)


No anexo abaixo, esta a forma com que faço o uso da LIBeSocial com o arquivo que gero(exemplo acima). Caso esteja fazendo o uso da forma errada, poderia estar me apontando para eu realizar a correção.

![image](https://github.com/qualitaocupacional/libesocial/assets/72530694/7c83cd3f-e836-44e1-a103-2143f299984c)

Desde já agradeço o retorno e fico no aguardo

**Reply by andidevel** [reply]
Olá @viniciusfeld, valeu pelas info.

> 
> [envio_remun.zip](https://github.com/qualitaocupacional/libesocial/files/12090548/envio_remun.zip)
> 

O XML do evento parece OK. Eu esperava que tivesse o XML de retorno do eSocial com os erros.

Tipo aqui:
> 
> ![image](https://user-images.githubusercontent.com/72530694/253942830-068662ec-f862-45cc-a64f-21827cfd0216.png)
> 

Esse print mostra que a resposta do **eSocial** ao seu lote de eventos tem erros. Se você quiser recuperar o retorno desse envio para saber o que deu problema, não precisa enviar o evento novamente. Imagino que você tenha armazenado o código do protocolo de envio, então de posse desse código, você pode fazer um script:
```python
import esocial.xml
import esocial.client
import json

ide_empregador = {
    'tpInsc': 1,
    'nrInsc': '12345678901234' # CNPJ/CPF completo (com 14/11 dígitos)
}

ide_transmissor = {
    'tpInsc': 1,
    'nrInsc': '43210987654321' # CNPJ/CPF completo (com 14/11 dígitos)
}

esocial_ws = esocial.client.WSClient(
    pfx_file='caminho/para/o/arquivo/certificado/A1',
    pfx_passw='senha do arquivo de certificado',
    employer_id=ide_empregador,
    # Se o transmissor é o próprio empregador, não precisa informar o "sender_id"
    sender_id=ide_transmissor,
)

# De posse do número do protocolo de envio
response = esocial_ws.retrieve('1.2.202109.0000000000000000001')

response_decoded = esocial.xml.decode_response(response)
print(json.dumps(response_decoded.toDict(), indent=4))
```
Você vai ter uma resposta mais ou menos assim:
```json
{
    "status": {
        "ocorrencias": [],
        "cdResposta": "201",
        "descResposta": "Lote processado com sucesso."
    },
    "lote": {
        "dhRecepcao": "2021-10-04T11:45:44.16",
        "versaoAplicativoRecepcao": "0.1.105",
        "protocoloEnvio": "1.1.202110.0000000000011111111"
    },
    "eventos": [
        {
            "id": "ID1123456780000002021100411454300001",
            "processamento": {
                "ocorrencias": [
                    {
                        "codigo": "xxx",
                        "descricao": "...",
                        "tipo": "1",
                        "localizacao": "..."
                    }
                ],
                "cdResposta": "401",
                "descResposta": "Conteudo do evento inválido.",
                "versaoAppProcessamento": "13.3.1",
                "dhProcessamento": "2021-10-04T11:45:50.923"
            },
            "recibo": null
        },
    ]
}
```
Na chave ```processamento``` -> ```ocorrencias``` de cada evento do lote, vai ter um ou mais dicts ```{"codigo":"xxx", "descricao":"...", "tipo":"1 ou 2", "localizacao":""}``` contendo o código, descrição e tipo do erro. Se o tipo for **1** é um erro, se for **2** é uma advertência [1].

Com essa informação você pode ter uma idéia de qual foi o problema com o envio do seu evento.

[1] [Manual de Orientação do Desenvolvedor eSocial](https://www.gov.br/esocial/pt-br/documentacao-tecnica/manuais/manualorientacaodesenvolvedoresocialv1-12.pdf)

**Reply by viniciusfeld** [reply]
Olá @andidevel, obrigado pelo retorno!

Segue a resposta quando consulto o protocolo de envio:

    "status": {
        "ocorrencias": [],
        "cdResposta": "201",
        "descResposta": "Lote processado com sucesso."
    },
    "lote": {
        "dhRecepcao": "2023-07-19T23:06:36.63",
        "versaoAplicativoRecepcao": "0.1.0.0",
        "protocoloEnvio": "1.1.202307.0000000005162812114"
    },
    "eventos": [
        {
            "id": "ID1121991210000002023071923061800001",
            "processamento": {
                "ocorrencias": [
                    {
                        "tipo": "1",
                        "codigo": "459",
                        "descricao": "N\u00e3o foi localizado um evento para o recibo de entrega informado ou o mesmo foi exclu\u00eddo/retificado.\nA\u00e7\u00e3o Sugerida: Dever\u00e1 existir um Evento j\u00e1 recebido, \"Ativo\" (n\u00e3o exclu\u00eddo ou retificado), com n\u00famero de recibo de entrega igual ao informado no campo.",
                        "localizacao": null
                    }
                ],
                "cdResposta": "401",
                "descResposta": "Conteudo do evento inv\u00e1lido.",
                "versaoAppProcessamento": "14.6.1",
                "dhProcessamento": "2023-07-19T23:06:36.703"
            },
            "recibo": null
        }
    ]
}

Este número de recibo está correto, pois quando faço a consulta no eSocial diretamente, ele me apresenta o mesmo recibo.

Também fiz o teste com outros funcionarios, e tive o mesmo problema de recibo.

Desde já agradeço o retorno e fico no aguardo.

**Reply by andidevel** [reply]
Olá @viniciusfeld,

Então, não me parece um problema com a **LIBeSocial**. No manual de orientação do **eSocial**, páginas 41-42:
"(...) Se o evento S-1299 já foi enviado, encerrando o movimento para determinado período de 
apuração, em caso de qualquer retificação no grupo de eventos periódicos S-1200 a S-1270, para 
aquele período de apuração, exceto para o S-1210, o respectivo movimento deve ser reaberto 
utilizando-se o evento S-1298, possibilitando o envio de retificações."

Sugiro dar uma olhada no manual de orientaçâo geral/desenvolvedor para ver os precedimentos no seu caso. No meu caso eu usava a lib para envio de eventos SST, não sei muito sobre os outros eventos.

**Comment by viniciusfeld** [comment]
Olá @andidevel, perdão pela demora no retorno.

Queria informar que consegui fazer o envio dos eventos de remuneração e de pagamento utilizando a LIBeSocial. Só uma duvida, vc recomenda usar a LIB que esta na master, ou na development? Pois vi que a development tem commits mais recentes.

Quero agradecer toda ajuda e suporte que você prestou nesses ultimos dias, foram de muita valia para o sucesso no envio, sua LIB ta show demais!!

**Reply by andidevel** [reply]
Olá @viniciusfeld, tudo certo?

> Queria informar que consegui fazer o envio dos eventos de remuneração e de pagamento utilizando a LIBeSocial. Só uma duvida, vc recomenda usar a LIB que esta na master, ou na development? Pois vi que a development tem commits mais recentes.
> 

Fico feliz que você conseguiu enviar os eventos. Quanto a qual branch utilizar, se a **main** está funcionando legal com você, continue com ela. Se você começar a ter problemas por causa de versões mais novas dos layouts do **eSocial**, daí você pode usar a **develop**.  Só não fiz merge ainda da **develop** na **main** pois queria adicionar a possibilidade de atualizar os XSD mais novos do **eSocial**, via CLI, sem precisar atualizar a lib inteira, mas isso ainda está em desenvolvimento, e estou fazendo uns *refactors*, 😃 .
